lib/instrumentation/bunyan.js

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const {
+  isApplicationLoggingEnabled,
+  isLogForwardingEnabled,
+  isLocalDecoratingEnabled,
+  isMetricsEnabled,
+  createModuleUsageMetric,
+  incrementLoggingLinesMetrics
+} = require('../util/application-logging')
+
+const MAX_LENGTH = 1021
+const OUTPUT_LENGTH = 1024
+const truncate = (str) => {
+  if (str && str.length > OUTPUT_LENGTH) {
+    return str.substring(0, MAX_LENGTH) + '...'
+  }
+
+  return str
+}
+
+const logger = require('../logger').child({ component: 'bunyan' })
+
+module.exports = function instrument(agent, bunyan, _, shim) {
+  const config = agent.config
+
+  if (!isApplicationLoggingEnabled(config)) {
+    logger.debug('Application logging not enabled. Not instrumenting bunyan.')
+    return
+  }
+
+  const logForwardingEnabled = isLogForwardingEnabled(config, agent)
+  const metricsEnabled = isMetricsEnabled(config)
+
+  if (logForwardingEnabled || metricsEnabled) {
+    shim.wrap(bunyan, 'createLogger', function wrapCreateLogger(_shim, createLogger) {
+      return function wrappedCreateLogger() {
+        createModuleUsageMetric('bunyan', agent.metrics)
+
+        const bunyanLogger = createLogger.apply(this, arguments)
+
+        if (logForwardingEnabled) {
+          bunyanLogger.addStream({
+            type: 'raw',
+            level: bunyanLogger.level(),
+            stream: {
+              write: function nrLogWrite(logLine) {
+                agent.logs.add(logLine)
+              }
+            }
+          })
+        }
+        return bunyanLogger
+      }
+    })
+  }
+
+  const localDecoratingEnabled = isLocalDecoratingEnabled(config)
+
+  shim.wrap(bunyan.prototype, '_emit', function wrapEmit(shim, emit) {
+    return function wrappedEmit() {
+      const args = shim.argsToArray.apply(shim, arguments)
+      const rec = args[0] || {}
+
+      if (metricsEnabled) {
+        incrementLoggingLinesMetrics(bunyan.nameFromLevel[rec.level], agent.metrics)
+      }
+
+      if (!localDecoratingEnabled && !logForwardingEnabled) {
+        return emit.apply(this, arguments)
+      }
+
+      // timestamp needs to be milliseconds since epoch
+      if (rec.timestamp) {
+        logger.traceOnce(
+          'Overwriting `timestamp` key; assigning original value to `original_timestamp`.'
+        )
+        rec.original_timestamp = rec.timestamp
+      }
+      rec.timestamp = Date.now()
+
+      // put log message into a consistent spot and ensure it's not too long
+      rec.message = truncate(rec.msg)
```

Comment:
```suggestion
if (isLocalDecoratingEnabled) {
      rec.message = rec.msg + agent.getNRLinkingMetadata()
   }
```

Reply:
Your suggested change would result in the `NR-LINKING` piped metadata being embedded inside the JSON object that is logged. Is that OK? Or does it need to be at the end just before the new-line character? Bunyan makes that quite difficult because it handles all the JSON stringification internally with non-exported (aka non-monkey-patchable) functions. The only viable option (that I can think of) would be to wrap the `write` method of each of the streams which is the point where we have a string that we can append to.

Bottom line question: is it an option to modify the fields of the log message when local decorating is enabled or should I go down the route of wrapping all the `stream.write` methods to inject the `NR-LINKING` piped metadata at that layer?

Reply:
> Your suggested change would result in the NR-LINKING piped metadata being embedded inside the JSON object that is logged. Is that OK?

Yes that is the spec for this feature of local log decorating.  The log forwarding feature operates on a copy of the log line which is why I suggest putting the rest of this logic in the stream.write function.  So you can see [here](https://github.com/newrelic/node-newrelic/blob/main/lib/instrumentation/winston.js#L86) we edit the existing log line for local log decorating and use it to increment metrics.  Then [here](https://github.com/newrelic/node-newrelic/blob/main/lib/instrumentation/winston.js#L46) we do the rest of the  logic in a winston formatter which does not affect the original log line